R/runCalibration.R

```diff
@@ -0,0 +1,1094 @@
+#' @param p_renovationCalibTarget data frame of calibration target for renovation flows
+#' @param p_stockCalibTarget data frame of calibration target for the stock, optional
+#'
+.addTargetsToInput <- function(mInput, path, p_constructionCalibTarget, p_renovationCalibTarget,
+                               p_stockCalibTarget = NULL) {
+  if (!is.null(p_stockCalibTarget)) {
+    invisible(mInput$addParameter(
+      name = "p_stockCalibTarget",
+      domain = c("qty", "bs", "hs", "vin", "reg", "loc", "typ", "inc", "ttot"),
+      records = rename(p_stockCalibTarget, value = "target"),
+      description = "historic stock of buildings as calibration target in million m2"
+    ))
+  }
+
+  invisible(mInput$addParameter(
+    name = "p_constructionCalibTarget",
+    domain = c("qty", "bs", "hs", "reg", "loc", "typ", "inc", "ttot"),
+    records = rename(p_constructionCalibTarget, value = "target"),
+    description = "historic flow of new buildings as calibration target in million m2/yr"
+  ))
+
+  invisible(mInput$addParameter(
+    name = "p_renovationCalibTarget",
+    domain = c("qty", "bs", "hs", "bsr", "hsr", "vin", "reg", "loc", "typ", "inc", "ttot"),
+    records = rename(p_renovationCalibTarget, value = "target"),
+    description = "historic flow of renovated and untouched buildings as calibration target in million m2/yr"
+  ))
+
+  mInput$write(file.path(path, "input.gdx"), compress = TRUE)
+
+}
+
+#' Compute the deviation between historic data and Brick results.
+#' Then compute the adjustment term for the calibration \code{d}
+#'
+#' @param m Gams transfer container with current Brick results
+#' @param target data frame of historical data
+#' @param dims character, dimensions of historic data and Brick results
+#' @param tcalib numeric, time steps to calibrate on
+#' @param flow character, either 'construction' or 'renovation'
+#' @param renAllowed data frame with allowed renovation transitions
+#'
+#' @importFrom dplyr %>% .data case_match case_when filter left_join mutate right_join select
+#' @importFrom tidyr pivot_wider replace_na
+#'
+.computeDeviation <- function(m, target, dims, tcalib, flow = "construction", renAllowed = NULL) {
+
+  varSym <- if (flow == "construction") "v_construction" else "v_renovation"
+  costSym <- if (flow == "construction") "p_specCostCon" else "p_specCostRen"
+
+  gamsVar <- readSymbol(m, symbol = varSym)
+
+  if (flow != "construction") {
+    gamsVar <- right_join(gamsVar, renAllowed, by = c("bs", "hs", "bsr", "hsr"))
+  }
+
+  cost <- readSymbol(m, symbol = costSym) %>%
+    pivot_wider(names_from = .data[["cost"]]) %>%
+    replace_na(list(intangible = 0)) # Set missing intangible costs to 0
+
+  do.call(expandSets, c(as.list(dims), .m = m)) %>%
+    left_join(gamsVar, by = dims) %>%
+    left_join(target, by = dims) %>%
+    left_join(cost, by = dims) %>%
+    replace_na(list(value = 0, target = 0)) %>%
+    filter(.data[["ttot"]] %in% tcalib) %>%
```

Comment:
Likely completely irrelevant. But this order should be more efficient ;)
```suggestion
    filter(.data[["ttot"]] %in% tcalib) %>%
    replace_na(list(value = 0, target = 0)) %>%
```